teps/000X-trigger-crd.md

```diff
@@ -0,0 +1,366 @@
+100 of projects, we would need 100 of pods just for acting as sink for these webhook events like github.
+These pods in turn consume resources.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+1. Reduction in resource cosumption due to EventListener.
+2. Ability to have single EventListener to cater to whole cluster.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+How to create Multitenant EventListener resource with appropriate permissions won't be part of this proposal.
+This will be handled by operator or admin of the cluster.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+It must be possible for EventListener to function in Multitenant scenarios. A single EventListener 
+should be able to process events for triggertemplate resources in different namespaces.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+EventListener CRD will be split into two - a Trigger CRD and a EventListener CRD.
+
+### Trigger CRD
+A single Trigger defines configuration for processing events i.e. it consists of TriggerBindings,
+a single TriggerTemplate, and optionally interceptors and a serviceAccount. 
+```
+apiVersion: v1alpha1
+kind: Trigger
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
+  interceptors:
+    cel:
+      filter: "$(header.eventType == "push")"
+  bindings:
+    - name: key
+      value: "$(body.commit_id)"
+  template: "ref-to-my-template"
+ status:
+   address:
+      url: "el-my-svc.cluster.local" # could also be an IP address?
+   conditions:
+      status: True
+      message: "Bound to EventListener my-el"
+```
+
+#### Trigger Status
+Trigger status will report url of EventListener bound and EventListener information to 
+which it is bound.
+
+### EventListener CRD
+EventListeners expose an addressable "Sink" to which incoming events in the form of HTTP requests
+are directed and process those requests with configuration from bound Triggers. Boundnamespace acts 
+as selector based on namespace. All triggers in namespaces specified are selected can be served by
+EventListener.
+
+```
+apiVersion: v1alpha1
+kind: EventListener
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
+  boundnamespace:
+      - foo-ns
+      - bar-ns
+```
+
+### Path based EventListener
+Inside EventListener, we will deduce Trigger to execute based on the path of the request URL.
+To refer to trigger resource foo in namespace bar, we would url ```/bar/foo```. Instead of 
```

Comment:
In this model, is there a way to serve multiple Triggers from the same webhook without having to deliver the webhook twice?

Reply:
No. I don't think so.

Reply:
Hmm, so this is a breaking change in how Triggers works today i.e. the lack of support of multiple triggers from the same webhook event. We should highlight this and document possible alternatives/workarounds.

- one idea: Maybe one Trigger CRD can contain multiple triggers?

Reply:
I found a workaround this.Pass label information via url to select TriggerCRD.

Reply:
That's an interesting alternative....we can try out an example maybe like creating a GitHub webhook with query params and see if that works?

Reply:
If the labels change, we'd have to reconfigure the webhook.

Reply:
I tested Github Webhook with query params and it works with the query params.

Reply:
Thanks, query params seem like an interesting idea though I still think it might be worth exploring if we can put multiple configs in the same Trigger or other alternatives which do not require label manipulation. I do think that we can work on this iteratively i.e we can get this TEP merged as proposed and then work on this part of the design more?

Reply:
Yes, that works for me.